src/dbnode/persist/fs/cross_block_reader.go

```diff
@@ -0,0 +1,168 @@
+	id                     ident.ID
+	tags                   ident.TagIterator
+	data                   checked.Bytes
+	checksum               uint32
+}
+
+type minHeap []*minHeapEntry
+
+// NewCrossBlockReader constructs a new CrossBlockReader based on given DataFileSetReaders.
+// DataFileSetReaders must be configured to return the data in the order of index, and must be
+// provided in a slice sorted by block start time.
+// Callers are responsible for closing the DataFileSetReaders.
+func NewCrossBlockReader(dataFileSetReaders []DataFileSetReader) (*crossBlockReader, error) {
+	var previousStart time.Time
+	for _, dataFileSetReader := range dataFileSetReaders {
+		if !dataFileSetReader.IsOrderedByIndex() {
+			return nil, errReaderNotOrderedByIndex
+		}
+		currentStart := dataFileSetReader.Range().Start
+		if !currentStart.After(previousStart) {
+			return nil, fmt.Errorf("dataFileSetReaders are not ordered by time (%s followed by %s)", previousStart, currentStart)
+		}
+		previousStart = currentStart
+	}
+
+	return &crossBlockReader{dataFileSetReaders: dataFileSetReaders}, nil
+}
+
+func (r *crossBlockReader) Read() (id ident.ID, tags ident.TagIterator, data checked.Bytes, checksum uint32, err error) {
+	if !r.initialized {
+		r.initialized = true
+		err := r.init()
+		if err != nil {
+			return nil, nil, nil, 0, err
+		}
+	}
+
+	if len(r.minHeap) == 0 {
+		return nil, nil, nil, 0, io.EOF
+	}
+
+	entry := heap.Pop(&r.minHeap).(*minHeapEntry)
+	if r.dataFileSetReaders[entry.dataFileSetReaderIndex] != nil {
+		nextEntry, err := r.readFromDataFileSet(entry.dataFileSetReaderIndex)
+		if err == io.EOF {
+			// will no longer read from this one
+			r.dataFileSetReaders[entry.dataFileSetReaderIndex] = nil
+		} else if err != nil {
+			return nil, nil, nil, 0, err
+		} else {
+			heap.Push(&r.minHeap, nextEntry)
+		}
+	}
+
+	return entry.id, entry.tags, entry.data, entry.checksum, nil
+}
+
+func (r *crossBlockReader) init() error {
+	r.minHeap = make([]*minHeapEntry, 0, len(r.dataFileSetReaders))
+
+	for i := range r.dataFileSetReaders {
+		entry, err := r.readFromDataFileSet(i)
+		if err == io.EOF {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		r.minHeap = append(r.minHeap, entry)
+	}
+
+	heap.Init(&r.minHeap)
+
+	return nil
+}
+
+func (r *crossBlockReader) readFromDataFileSet(index int) (*minHeapEntry, error) {
+	id, tags, data, checksum, err := r.dataFileSetReaders[index].Read()
+
+	if err == io.EOF {
+		return nil, err
+	}
+
+	if err != nil {
+		closeErr := r.Close()
+		if closeErr != nil {
+			return nil, xerrors.NewMultiError().Add(err).Add(closeErr)
+		}
+		return nil, err
+	}
+
+	return &minHeapEntry{
+		dataFileSetReaderIndex: index,
+		id:                     id,
+		tags:                   tags,
+		data:                   data,
+		checksum:               checksum,
+	}, nil
+}
+
+func (r *crossBlockReader) Close() error {
+	for _, entry := range r.minHeap {
+		entry.id.Finalize()
+		entry.tags.Close()
+		entry.data.DecRef()
```

Comment:
Don't we already `defer DecRef()` inside of the `DataFileSetReader` already before returning the underlying data? 

```
var data checked.Bytes
if r.bytesPool != nil {
        data = r.bytesPool.Get(int(entry.Size))
        data.IncRef()
        defer data.DecRef()
} else {
        data = checked.NewBytes(make([]byte, 0, entry.Size), nil)
        data.IncRef()
        defer data.DecRef()
}
```

I also commented this on the other PR: https://github.com/m3db/m3/pull/2465#discussion_r460284933

Reply:
Hmm, this needs some additional thought, I guess I'll wait for @robskillington feedback on the questions that you raised on https://github.com/m3db/m3/pull/2465.